### PR TITLE
Add first draft for finalize() method

### PR DESCRIPTION
**Pull request by AbrilRBS**
For https://github.com/conan-io/conan/pull/16646

### CODE REVIEW

File: reference/conanfile/methods.rst [hunk 1]
```diff
@@ -43,6 +44,7 @@ What follows is a list of methods that you can define in your recipes to customi
 - :doc:`deploy() <methods/deploy>`: Deploys (copy from package to user folder) the desired artifacts
 - :doc:`export() <methods/export>`: Copies files that are part of the recipe
 - :doc:`export_sources() <methods/export_sources>`: Copies files that are part of the recipe sources
+- :doc:`finalize() <methods/finalize>`: Customizes the package for using it in the running machine without affecting the original package
```
Comment (AbrilRBS) [inline reference/conanfile/methods.rst:47]: Or something more in line with "Allows to isolate user-specific configurations..."?

File: reference/conanfile/methods/finalize.rst [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+------------------------
+
+The most common use-case of this method is to avoid corrupting the immutability of the package.
+
+For example, if a package generates some files during its execution, like a python package that generates pycache files,
+you can use this method to generate those files in the local machine, without affecting the original package.
+This is the case for tools like Meson.
+
+
+.. code-block:: python
```
Comment (AbrilRBS) [inline reference/conanfile/methods/finalize.rst:38]: This is a bit barren, I'd like to have a more complete example but don't know how much further to take it

File: reference/conanfile/methods/finalize.rst [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+finalize()
+==========
+
+.. include:: ../../../common/experimental_warning.inc
+
+Package immutability is an important concept in Conan. It ensures that the package is not modified after it has been built and packaged,
+so that the package id is consistent and the package can be reused in different machines.
+
+This method is intended for customization of the package in the running machine, allowing modifications to the package
```
Comment (AbrilRBS) [inline reference/conanfile/methods/finalize.rst:11]: I'd say this would need a bit more explanation/setup for the problem/approach to fix it, wdyt?

**Review by AbrilRBS — COMMENTED** [review]
Some comments

File: reference/conanfile/methods.rst (hunk 1 above)
Comment (memsharded) [reply reference/conanfile/methods.rst:47]: The current line looks good to me.

File: reference/conanfile/methods/finalize.rst [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+so that any modifications done by the package are done in the local finalized folder, without affecting the original package.
+
+For cases where it's necessary to access the original package, the ``immutable_package_folder`` attribute is available
+both in the same recipe's ``self.immutable_package_folder`` and
+thru the ``self.dependencies[<package_name>].immutable_package_folder`` attribute in the dependants' recipe.
+This info is also serialized as part of the graph information in :command:`conan graph info` etc.
+
+.. note::
+
+  Without using this approach, the package would generate the pycache files in the package folder,
```
Comment (memsharded) [inline reference/conanfile/methods/finalize.rst:66]: Clarify that this can raise an error if ``conan upload --check`` or ``conan cache check-integrity``

**Review by memsharded — APPROVED** [review]

**Pull request merged**